File: src/hayulo/intent.py
from __future__ import annotations

import re
from typing import Any

from .diagnostics import Diagnostic, HayuloSyntaxError


def parse_top_level_intent(source: str, filename: str | None = None) -> dict[str, Any] | None:
    lines = source.splitlines()
    intent: dict[str, Any] | None = None
    depth = 0
    i = 0

    while i < len(lines):
        line = _strip_comment(lines[i])
        if depth == 0 and re.match(r"\s*intent\s*\{", line):
            block, i = _collect_block(lines, i, filename, "intent")
            values = _parse_fields(block, filename)
            if intent is None:
                intent = {}
            intent.update(values)
            continue

        depth = max(0, depth + _count_braces(line))
        i += 1

    return intent


def _collect_block(lines: list[str], start: int, filename: str | None, label: str) -> tuple[list[tuple[int, str]], int]:
    head = _strip_comment(lines[start])
    if "{" not in head:
        raise HayuloSyntaxError(Diagnostic(code="missing_block", message=f"Expected '{{' for {label} block.", file=filename, line=start + 1))

    after_open = head.split("{", 1)[1]
    depth = _count_braces(head)
    block: list[tuple[int, str]] = []

    if depth <= 0:
        before_close = after_open.rsplit("}", 1)[0].strip()
        if before_close:
            block.append((start + 1, before_close))
        return block, start + 1

    if after_open.strip():
        block.append((start + 1, after_open))

    i = start + 1
    while i < len(lines):
        delta = _count_braces(lines[i])
        if depth + delta <= 0:
            before_close = lines[i].rsplit("}", 1)[0].strip()
            if before_close:
                block.append((i + 1, before_close))
            return block, i + 1
        block.append((i + 1, lines[i]))
        depth += delta
        i += 1

    raise HayuloSyntaxError(Diagnostic(code="unterminated_block", message=f"Unterminated {label} block.", file=filename, line=start + 1, suggestions=["Add a closing '}'."]))


def _parse_fields(block: list[tuple[int, str]], filename: str | None) -> dict[str, Any]:
    values: dict[str, Any] = {}
    i = 0

    while i < len(block):
        line_no, raw = block[i]
        line = _clean_line(raw)
        if not line:
            i += 1
            continue

        match = re.match(r"([A-Za-z_]\w*)\s*:\s*(.+)$", line)
        if not match:
            _error("invalid_intent_field", f"Invalid intent field: {line!r}.", filename, line_no, ["Use: key: \"value\" or key: [\"value\"]."])

        key = match.group(1)
        rhs = match.group(2).strip()

        if rhs.startswith('"'):
            values[key] = _parse_quoted_scalar(rhs, filename, line_no)
            i += 1
            continue

        if rhs.startswith("["):
            parts = [rhs]
            bracket_depth = _count_brackets(rhs)
            i += 1
            while bracket_depth > 0 and i < len(block):
                next_line_no, next_raw = block[i]
                next_line = _clean_line(next_raw)
                parts.append(next_line)
                bracket_depth += _count_brackets(next_line)
                line_no = next_line_no
                i += 1
            if bracket_depth != 0:
                _error("unterminated_intent_list", f"Unterminated intent list for {key!r}.", filename, line_no, ["Add a closing ']'."])
            values[key] = _parse_string_list(" ".join(parts), filename, line_no)
            continue

        _error("invalid_intent_value", f"Invalid intent value for {key!r}.", filename, line_no, ["Use a string or list of strings."])

    return values


def _parse_quoted_scalar(text: str, filename: str | None, line: int) -> str:
    value, index = _consume_quoted(text, 0, filename, line)
    trailing = text[index:].strip()
    if trailing not in {"", ","}:
        _error("invalid_intent_value", "Intent string values cannot contain trailing tokens.", filename, line, ["Use a single quoted string value."])
    return value


def _parse_string_list(text: str, filename: str | None, line: int) -> list[str]:
    text = text.strip()
    if text.endswith("],"):
        text = text[:-1].rstrip()
    if not text.startswith("[") or not text.endswith("]"):
        _error("invalid_intent_value", "Intent list values must start with '[' and end with ']'.", filename, line, ["Use a list of quoted strings."])

    values: list[str] = []
    index = 1
    end = len(text) - 1
    while index < end:
        while index < end and text[index].isspace():
            index += 1
        if index < end and text[index] == ",":
            index += 1
            continue
        if index >= end:
            break
        if text[index] != '"':
            _error("invalid_intent_value", "Intent lists can only contain strings.", filename, line, ["Use quoted strings inside intent lists."])
        value, index = _consume_quoted(text, index, filename, line)
        values.append(value)
        while index < end and text[index].isspace():
            index += 1
        if index < end and text[index] == ",":
            index += 1
            continue
        if index < end:
            _error("invalid_intent_value", "Intent list items must be separated by commas.", filename, line, ["Add a comma between list items."])
    return values


def _consume_quoted(text: str, start: int, filename: str | None, line: int) -> tuple[str, int]:
    if start >= len(text) or text[start] != '"':
        _error("invalid_intent_value", "Expected quoted string.", filename, line, ["Use double quotes for intent strings."])

    chars: list[str] = []
    i = start + 1
    while i < len(text):
        ch = text[i]
        if ch == '"':
            return "".join(chars), i + 1
        if ch == "\\":
            i += 1
            if i >= len(text):
                break
            escaped = text[i]
            chars.append({"n": "\n", "t": "\t", '"': '"', "\\": "\\"}.get(escaped, escaped))
        else:
            chars.append(ch)
        i += 1

    _error("unterminated_string", "Unterminated intent string.", filename, line, ["Add a closing double quote."])


def _clean_line(line: str) -> str:
    return _strip_comment(line).strip()


def _strip_comment(line: str) -> str:
    in_string = False
    escaped = False
    i = 0
    while i < len(line):
        ch = line[i]
        if escaped:
            escaped = False
        elif ch == "\\":
            escaped = True
        elif ch == '"':
            in_string = not in_string
        elif not in_string and line[i : i + 2] == "//":
            return line[:i]
        i += 1
    return line


def _count_braces(line: str) -> int:
    return _count_unquoted(line, "{", "}")


def _count_brackets(line: str) -> int:
    return _count_unquoted(line, "[", "]")


def _count_unquoted(line: str, open_char: str, close_char: str) -> int:
    text = _strip_comment(line)
    in_string = False
    escaped = False
    count = 0
    for ch in text:
        if escaped:
            escaped = False
        elif ch == "\\":
            escaped = True
        elif ch == '"':
            in_string = not in_string
        elif not in_string and ch == open_char:
            count += 1
        elif not in_string and ch == close_char:
            count -= 1
    return count


def _error(code: str, message: str, filename: str | None, line: int, suggestions: list[str]) -> None:
    raise HayuloSyntaxError(Diagnostic(code=code, message=message, file=filename, line=line, suggestions=suggestions))

File: src/intent.rs
use std::collections::BTreeMap;

use regex::Regex;
use serde_json::{Value, json};

use crate::diagnostics::{Diagnostic, HayuloError, HayuloResult};
use crate::formatter::strip_comment;

pub fn parse_top_level_intent(source: &str, filename: Option<&str>) -> HayuloResult<Option<Value>> {
    let lines: Vec<&str> = source.lines().collect();
    let mut intent = BTreeMap::new();
    let mut depth = 0isize;
    let mut i = 0usize;
    let re = Regex::new(r"^\s*intent\s*\{").unwrap();

    while i < lines.len() {
        let line = strip_comment(lines[i]);
        if depth == 0 && re.is_match(&line) {
            let (block, next) = collect_block(&lines, i, filename, "intent")?;
            for (key, value) in parse_fields(&block, filename)? {
                intent.insert(key, value);
            }
            i = next;
            continue;
        }
        depth = (depth + count_braces(&line)).max(0);
        i += 1;
    }

    if intent.is_empty() {
        Ok(None)
    } else {
        Ok(Some(json!(intent)))
    }
}

fn collect_block(
    lines: &[&str],
    start: usize,
    filename: Option<&str>,
    label: &str,
) -> HayuloResult<(Vec<(usize, String)>, usize)> {
    let head = strip_comment(lines[start]);
    if !head.contains('{') {
        return Err(HayuloError::new(
            Diagnostic::new("missing_block", format!("Expected '{{' for {label} block."))
                .file(filename)
                .line(Some(start + 1)),
        ));
    }
    let after_open = head.split_once('{').map(|(_, right)| right).unwrap_or("");
    let mut depth = count_braces(&head);
    let mut block = Vec::new();
    if depth <= 0 {
        let before_close = after_open
            .rsplit_once('}')
            .map(|(left, _)| left)
            .unwrap_or(after_open)
            .trim();
        if !before_close.is_empty() {
            block.push((start + 1, before_close.to_string()));
        }
        return Ok((block, start + 1));
    }
    if !after_open.trim().is_empty() {
        block.push((start + 1, after_open.to_string()));
    }
    let mut i = start + 1;
    while i < lines.len() {
        let delta = count_braces(lines[i]);
        if depth + delta <= 0 {
            let before_close = lines[i]
                .rsplit_once('}')
                .map(|(left, _)| left)
                .unwrap_or(lines[i])
                .trim();
            if !before_close.is_empty() {
                block.push((i + 1, before_close.to_string()));
            }
            return Ok((block, i + 1));
        }
        block.push((i + 1, lines[i].to_string()));
        depth += delta;
        i += 1;
    }
    Err(HayuloError::new(
        Diagnostic::new("unterminated_block", format!("Unterminated {label} block."))
            .file(filename)
            .line(Some(start + 1))
            .suggestion("Add a closing '}'."),
    ))
}

fn parse_fields(
    block: &[(usize, String)],
    filename: Option<&str>,
) -> HayuloResult<Vec<(String, Value)>> {
    let mut values = Vec::new();
    let mut i = 0usize;
    let field_re = Regex::new(r"^([A-Za-z_]\w*)\s*:\s*(.+)$").unwrap();
    while i < block.len() {
        let (line_no, raw) = &block[i];
        let line = clean_line(raw);
        if line.is_empty() {
            i += 1;
            continue;
        }
        let captures = field_re.captures(&line).ok_or_else(|| {
            HayuloError::new(
                Diagnostic::new(
                    "invalid_intent_field",
                    format!("Invalid intent field: {line:?}."),
                )
                .file(filename)
                .line(Some(*line_no))
                .suggestion("Use: key: \"value\" or key: [\"value\"]."),
            )
        })?;
        let key = captures[1].to_string();
        let rhs = captures[2].trim();
        if rhs.starts_with('"') {
            values.push((key, json!(parse_quoted_scalar(rhs, filename, *line_no)?)));
            i += 1;
            continue;
        }
        if rhs.starts_with('[') {
            let mut parts = vec![rhs.to_string()];
            let mut bracket_depth = count_brackets(rhs);
            i += 1;
            let mut last_line = *line_no;
            while bracket_depth > 0 && i < block.len() {
                let (next_line_no, next_raw) = &block[i];
                let next_line = clean_line(next_raw);
                parts.push(next_line.clone());
                bracket_depth += count_brackets(&next_line);
                last_line = *next_line_no;
                i += 1;
            }
            if bracket_depth != 0 {
                return Err(HayuloError::new(
                    Diagnostic::new(
                        "unterminated_intent_list",
                        format!("Unterminated intent list for {key:?}."),
                    )
                    .file(filename)
                    .line(Some(last_line))
                    .suggestion("Add a closing ']'."),
                ));
            }
            values.push((
                key,
                json!(parse_string_list(&parts.join(" "), filename, last_line)?),
            ));
            continue;
        }
        return Err(HayuloError::new(
            Diagnostic::new(
                "invalid_intent_value",
                format!("Invalid intent value for {key:?}."),
            )
            .file(filename)
            .line(Some(*line_no))
            .suggestion("Use a string or list of strings."),
        ));
    }
    Ok(values)
}

fn parse_quoted_scalar(text: &str, filename: Option<&str>, line: usize) -> HayuloResult<String> {
    let (value, index) = consume_quoted(text, 0, filename, line)?;
    let trailing = text[index..].trim();
    if !trailing.is_empty() && trailing != "," {
        return Err(HayuloError::new(
            Diagnostic::new(
                "invalid_intent_value",
                "Intent string values cannot contain trailing tokens.",
            )
            .file(filename)
            .line(Some(line))
            .suggestion("Use a single quoted string value."),
        ));
    }
    Ok(value)
}

fn parse_string_list(text: &str, filename: Option<&str>, line: usize) -> HayuloResult<Vec<String>> {
    let mut text = text.trim().to_string();
    if text.ends_with("],") {
        text.pop();
        text = text.trim_end().to_string();
    }
    if !text.starts_with('[') || !text.ends_with(']') {
        return Err(HayuloError::new(
            Diagnostic::new(
                "invalid_intent_value",
                "Intent list values must start with '[' and end with ']'.",
            )
            .file(filename)
            .line(Some(line))
            .suggestion("Use a list of quoted strings."),
        ));
    }
    let mut values = Vec::new();
    let mut index = 1usize;
    let end = text.len() - 1;
    while index < end {
        while index < end && text.as_bytes()[index].is_ascii_whitespace() {
            index += 1;
        }
        if index < end && text.as_bytes()[index] == b',' {
            index += 1;
            continue;
        }
        if index >= end {
            break;
        }
        if text.as_bytes()[index] != b'"' {
            return Err(HayuloError::new(
                Diagnostic::new(
                    "invalid_intent_value",
                    "Intent lists can only contain strings.",
                )
                .file(filename)
                .line(Some(line))
                .suggestion("Use quoted strings inside intent lists."),
            ));
        }
        let (value, next) = consume_quoted(&text, index, filename, line)?;
        values.push(value);
        index = next;
        while index < end && text.as_bytes()[index].is_ascii_whitespace() {
            index += 1;
        }
        if index < end && text.as_bytes()[index] == b',' {
            index += 1;
            continue;
        }
        if index < end {
            return Err(HayuloError::new(
                Diagnostic::new(
                    "invalid_intent_value",
                    "Intent list items must be separated by commas.",
                )
                .file(filename)
                .line(Some(line))
                .suggestion("Add a comma between list items."),
            ));
        }
    }
    Ok(values)
}

fn consume_quoted(
    text: &str,
    start: usize,
    filename: Option<&str>,
    line: usize,
) -> HayuloResult<(String, usize)> {
    let chars: Vec<char> = text.chars().collect();
    if chars.get(start) != Some(&'"') {
        return Err(HayuloError::new(
            Diagnostic::new("invalid_intent_value", "Expected quoted string.")
                .file(filename)
                .line(Some(line))
                .suggestion("Use double quotes for intent strings."),
        ));
    }
    let mut value = String::new();
    let mut i = start + 1;
    while i < chars.len() {
        let ch = chars[i];
        if ch == '"' {
            let byte_index = chars[..=i].iter().map(|c| c.len_utf8()).sum();
            return Ok((value, byte_index));
        }
        if ch == '\\' {
            i += 1;
            if i >= chars.len() {
                break;
            }
            value.push(match chars[i] {
                'n' => '\n',
                't' => '\t',
                '"' => '"',
                '\\' => '\\',
                other => other,
            });
        } else {
            value.push(ch);
        }
        i += 1;
    }
    Err(HayuloError::new(
        Diagnostic::new("unterminated_string", "Unterminated intent string.")
            .file(filename)
            .line(Some(line))
            .suggestion("Add a closing double quote."),
    ))
}

fn clean_line(line: &str) -> String {
    strip_comment(line).trim().to_string()
}

fn count_braces(line: &str) -> isize {
    count_unquoted(line, '{', '}')
}

fn count_brackets(line: &str) -> isize {
    count_unquoted(line, '[', ']')
}

fn count_unquoted(line: &str, open: char, close: char) -> isize {
    let text = strip_comment(line);
    let mut in_string = false;
    let mut escaped = false;
    let mut count = 0isize;
    for ch in text.chars() {
        if escaped {
            escaped = false;
        } else if ch == '\\' {
            escaped = true;
        } else if ch == '"' {
            in_string = !in_string;
        } else if !in_string && ch == open {
            count += 1;
        } else if !in_string && ch == close {
            count -= 1;
        }
    }
    count
}
